core/trino-parser/src/main/java/io/trino/sql/tree/SetTimeZone.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.tree;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.trino.sql.tree.IntervalLiteral.IntervalField.HOUR;
+import static io.trino.sql.tree.IntervalLiteral.IntervalField.MINUTE;
+import static java.util.Objects.requireNonNull;
+
+public class SetTimeZone
+        extends Statement
+{
+    private final Optional<Expression> timeZone;
+
+    public SetTimeZone(Optional<Expression> timeZone)
+    {
+        this(Optional.empty(), timeZone);
+    }
+
+    public SetTimeZone(NodeLocation location, Optional<Expression> timeZone)
+    {
+        this(Optional.of(location), timeZone);
+    }
+
+    private SetTimeZone(Optional<NodeLocation> location, Optional<Expression> timeZone)
+    {
+        super(location);
+        requireNonNull(timeZone, "timeZone is null");
+        timeZone.ifPresent(expression -> {
+            checkArgument(expression instanceof IntervalLiteral || expression instanceof StringLiteral,
+                    "timeZone must be IntervalLiteral or StringLiteral");
+            if (expression instanceof IntervalLiteral) {
+                IntervalLiteral intervalLiteral = (IntervalLiteral) expression;
+                checkArgument(intervalLiteral.getStartField() == HOUR, "timeZone interval start field must be HOUR");
+                intervalLiteral.getEndField()
+                        .ifPresent(endField -> checkArgument(endField == MINUTE, "timeZone interval end field must be MINUTE"));
+            }
```

Comment:
This should be a semantic check, not a syntactic check. Anything that produces a type of value INTERVAL DAY TO HOUR is legal, including function calls that return such types. I.e., there's no requirement or guarantee that this be an interval literal.

Reply:
Thanks for the heads'up.
I made the appropriate modifications to support expressions given to `SET TIME ZONE`.

```
trino> set time zone concat_ws('/', 'America', 'Los_Angeles');
SET TIME ZONE
trino> select current_timestamp;
                    _col0                    
---------------------------------------------
 2021-06-08 01:31:18.761 America/Los_Angeles 
(1 row)
```

```
trino> set time zone fdsafdsa;
Query 20210608_092131_00001_r8ubx failed: line 1:15: Column 'fdsafdsa' cannot be resolved
set time zone fdsafdsa
```

```
trino> SET TIME ZONE parse_duration('8h');
SET TIME ZONE
trino> SET TIME ZONE parse_duration('3601s');
Query 20210608_113400_00004_vfuxk failed: Invalid time zone offset interval: interval contains seconds
```


I've found that some of the exception messages are now misleading (due to the fact that the `expression` newly instroduced in the grammar of `SET TIME ZONE` statement are actually mainly used for `SELECT` statements).